### ISSUE TITLE
Presize ByteArrayOrdinalMap where possible

### ISSUE THREAD

**Issue by DanielThomas** [Netflix/hollow#307]
`ByteArrayOrdinalMap` resizing and the rehashing for that resize is a significant contributor when running a producer cycle.

In particular I noticed that `com.netflix.hollow.core.write.HollowTypeWriteState#restoreFrom` has the opportunity to call `populatedOrdinals.cardinality()` and know exactly how many ordinals are incoming:

https://github.com/Netflix/hollow/blob/f0126b04575a897b3e7ceefa07baa0f354b79683/hollow/src/main/java/com/netflix/hollow/core/write/HollowTypeWriteState.java#L302-L305

In our case:

```
DeploymentLocation - 84
DeploymentType - 2
String - 168236
Deployment - 28249
DeploymentImage - 8347
ModuleMetadata - 20339
ModuleDependency - 146534
Owner - 1763
DeploymentUrls - 28245
ListOfModuleDependencyList - 75692
Module - 83021
ListOfString - 1201
BuildMetadata - 20260
ModuleDependencyList - 70184725
ListOfModuleDependency - 70184725
ModuleType - 6
ModuleVersionIdentifier - 145690
Status - 1
```

Here's a quick and dirty sampling profile from YourKit:

<img width="1275" alt="screen shot 2018-10-17 at 3 39 03 pm" src="https://user-images.githubusercontent.com/1479220/47120250-d5a6ce80-d222-11e8-88c3-1b28b28c9aa8.png">

However the big win here would be presizing `com.netflix.hollow.core.write.HollowTypeWriteState#ordinalMap` at that point, because that's where the resizes really sting:

<img width="1892" alt="screen shot 2018-10-16 at 4 20 30 pm" src="https://user-images.githubusercontent.com/1479220/47178063-f298db80-d2ce-11e8-8f18-466c9375e210.png">

After optimizing commonly duplicated classes with interning and `__assigned_ordinal` it's even more of a bottleneck (less total time of course):

<img width="939" alt="screen shot 2018-10-19 at 3 17 06 pm" src="https://user-images.githubusercontent.com/1479220/47246380-40d2db00-d3b2-11e8-91cc-053a065ce366.png">

There's also some non-concurrent rehashing in:

`com.netflix.hollow.core.write.HollowTypeWriteState#HollowTypeWriteState`

**Comment by DanielThomas** [comment]
Relates to https://github.com/Netflix/hollow/issues/305

**Comment by PaulSandoz** [comment]
Fixed in #378
